node/pkg/watchers/solana/tx_for_addr.go

```diff
@@ -0,0 +1,183 @@
+// the Watcher object. This is our starting point for subsequent poll intervals. This is updated each time
+// we observe a transaction. By storing it in the Watcher object, we can continue where we left off on a watcher
+// restart (but not on a guardian restart).
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"slices"
+	"time"
+
+	"github.com/gagliardetto/solana-go"
+	"github.com/gagliardetto/solana-go/rpc"
+	"go.uber.org/zap"
+)
+
+const (
+	// MaxSignaturesPerQuery is the maximum number of transactions to be returned by `GetSignaturesForAddressWithOpts`.
+	// The max / default is 1000. You can set this to something smaller (like 5) to test pagination.
+	MaxSignaturesPerQuery = 1000
+
+	// NumGetBlockRetries is how many times we will try to query for a slot, allowing for skipped / missing slots.
+	NumGetBlockRetries = 25
+)
+
+// transactionProcessor is the entry point of the runnable that periodically queries for new Wormhole observations.
+// It uses the standard `DefaultPollDelay`, although the timing will vary based on query delays. Each interval, it
+// looks for new transactions involving the core contract by using the `GetSignaturesForAddressWithOpts` RPC call.
+// Any transactions that are detected are processed using the standard transaction processing code.
+// Note: This is a separate runnable so that query delays don't impact the standard block height reporting.
+func (s *SolanaWatcher) transactionProcessor(ctx context.Context) error {
+	// Initialize our starting point. If we already have a previous signature, that means there has been a watcher restart
+	// (rather than a guardian restart), so we want to preserve that value and start where we left off.
+	if s.pollPrevWormholeSignature.IsZero() {
+		var err error
+		s.pollPrevWormholeSignature, err = s.getPrevWormholeSignature()
+		if err != nil {
+			s.logger.Error("failed to get the last wormhole signature on start up", zap.Error(err))
+			s.errC <- err
+			return err
+		}
+	}
+
+	s.logger.Info("starting from previous wormhole signature", zap.Stringer("prevSig", s.pollPrevWormholeSignature))
+
+	timer := time.NewTicker(DefaultPollDelay)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-timer.C:
+			//nolint:contextcheck // Passed via the 's' object instead of as a parameter.
+			err := s.processNewTransactions()
+			if err != nil {
+				s.logger.Error("failed to get transactions", zap.Error(err))
+				s.errC <- err
+				return err
+			}
+		}
+	}
+}
+
+// getPrevWormholeSignature reads the most recent transaction involving the Wormhole core contract and returns it.
+func (s *SolanaWatcher) getPrevWormholeSignature() (solana.Signature, error) {
+	limit := int(1)
+	signatures, err := s.rpcClient.GetSignaturesForAddressWithOpts(s.ctx, s.contract, &rpc.GetSignaturesForAddressOpts{
+		Commitment: s.commitment,
+		Limit:      &limit,
+	})
+
+	if err != nil || len(signatures) == 0 {
```

Comment:
just noting that I *think* this would be okay even for a new core bridge deployment because the core bridge must at least have been initialized first, meaning there would realistically always be at least 1 most recent transaction.

Reply:
Yes, that was my assumption. Don't start up the watcher until after you initialize the core.